vpr/src/route/rr_graph_reader.cpp

```diff
@@ -47,809 +35,751 @@
+	T_SWITCH,
+	T_TIMING,
+	T_SIZING,
+	T_SEGMENT,
+	T_BLOCK_TYPE,
+	T_PIN_CLASS,
+	T_PIN,
+	T_GRID_LOC,
+	T_NODE,
+	T_EDGE,
+	T_METADATA,
+	T_META,
+	T_LOC,
+};
+static std::unordered_map<std::string, tag_type> tag_map = {
+	{"rr_graph", T_RR_GRAPH},
+	{"channels", T_CHANNELS},
+	{"switches", T_SWITCHES},
+	{"segments", T_SEGMENTS},
+	{"block_types", T_BLOCK_TYPES},
+	{"grid", T_GRID},
+	{"rr_nodes", T_RR_NODES},
+	{"rr_edges", T_RR_EDGES},
+	{"channel", T_CHANNEL},
+	{"x_list", T_X_LIST},
+	{"y_list", T_Y_LIST},
+	{"switch", T_SWITCH},
+	{"timing", T_TIMING},
+	{"sizing", T_SIZING},
+	{"segment", T_SEGMENT},
+	{"block_type", T_BLOCK_TYPE},
+	{"pin_class", T_PIN_CLASS},
+	{"pin", T_PIN},
+	{"grid_loc", T_GRID_LOC},
+	{"node", T_NODE},
+	{"edge", T_EDGE},
+	{"metadata", T_METADATA},
+	{"meta", T_META},
+	{"loc", T_LOC},
+};
+
+/* The parser stack to keep the current path of tags.
+ * It can have a static size, because there is no recursive nesting in the file format. */
+static tag_type parser_stack[64];
+static int parser_stack_top = -1;
+
+/* The lookup table. It maps the previous tag and current tag to the handler function.
+ * It is initialized using a class because C++ doesn't like array initialization using enums. */
+class ParseTableClass {
+public:
+	ParseTableClass(){
+		table[T_RR_GRAPH][T_CHANNELS] = consume_channels;
+		table[T_RR_GRAPH][T_SWITCHES] = consume_switches;
+		table[T_RR_GRAPH][T_SEGMENTS] = consume_segments;
+		table[T_RR_GRAPH][T_BLOCK_TYPES] = consume_block_types;
+		table[T_RR_GRAPH][T_GRID] = consume_grid;
+		table[T_RR_GRAPH][T_RR_NODES] = consume_rr_nodes;
+		table[T_RR_GRAPH][T_RR_EDGES] = consume_rr_edges;
+		table[T_CHANNELS][T_CHANNEL] = consume_channel;
+		table[T_CHANNELS][T_X_LIST] = consume_x_list;
+		table[T_CHANNELS][T_Y_LIST] = consume_y_list;
+		table[T_SWITCHES][T_SWITCH] = consume_switch;
+		table[T_SWITCH][T_TIMING] = consume_switch_timing;
+		table[T_SWITCH][T_SIZING] = consume_switch_sizing;
+		table[T_SEGMENTS][T_SEGMENT] = consume_segment;
+		table[T_SEGMENT][T_TIMING] = consume_segment_timing;
+		table[T_BLOCK_TYPES][T_BLOCK_TYPE] = consume_block_type;
+		table[T_BLOCK_TYPE][T_PIN_CLASS] = consume_pin_class;
+		table[T_PIN_CLASS][T_PIN] = consume_pin;
+		table[T_GRID][T_GRID_LOC] = consume_grid_loc;
+		table[T_RR_NODES][T_NODE] = consume_node;
+		table[T_RR_EDGES][T_EDGE] = consume_edge;
+		table[T_NODE][T_LOC] = consume_node_loc;
+		table[T_NODE][T_TIMING] = consume_node_timing;
+		table[T_NODE][T_SEGMENT] = consume_node_segment;
+		table[T_NODE][T_METADATA] = consume_node_metadata;
+		table[T_EDGE][T_METADATA] = consume_edge_metadata;
+		table[T_METADATA][T_META] = consume_meta;
+	}
+	/* This works because the first [] is handled by this function,
+	 * returning an array of function pointers, which can be then regularly
+	 * accessed with the second []. */
+	void(**operator[](size_t index))(const char **) const {
+		return table[index];
+	}
+private:
+	void(*table[32][32])(const char **);
```

Comment:
Why 32? Magic numbers are bad :-P

Reply:
you are right. i thought the number of rr_graph tags would never exceed 32 and was too lazy to define an enum of `NUM_TAG_TYPES`.